components/script/script_task.rs

```diff
@@ -848,6 +893,76 @@ impl ScriptTask {
         true
     }
 
+    fn categorize_msg(&self, msg: &MixedMessage) -> ScriptTaskEventCategory {
+        match *msg {
+            MixedMessage::FromConstellation(ref inner_msg) => {
+                match *inner_msg {
+                    ConstellationControlMsg::SendEvent(_, _) =>
+                        ScriptTaskEventCategory::DomEvent,
+                    _ => ScriptTaskEventCategory::ConstellationMsg
+                }
+            },
+            MixedMessage::FromDevtools(_) => ScriptTaskEventCategory::DevtoolsMsg,
+            MixedMessage::FromImageCache(_) => ScriptTaskEventCategory::ImageCacheMsg,
+            MixedMessage::FromScript(ref inner_msg) => {
+                match *inner_msg {
+                    MainThreadScriptMsg::Common(CommonScriptMsg::RunnableMsg(ref category, _)) =>
+                        *category,
+                    _ => ScriptTaskEventCategory::ScriptEvent
+                }
+            }
+        }
+    }
+
+    fn profile_event<F, R>(&self, category: ScriptTaskEventCategory, f: F) -> R
+        where F: FnOnce() -> R {
+
+        if opts::get().profile_script_events {
+            let start = time::precise_time_ns();
+            let result = f();
+            let end = time::precise_time_ns();
+
+            let duration = end - start;
+
+            let aggregate = {
+                let zero = 0;
+                let perf_profiler_times = self.perf_profiler_times.borrow();
+                let so_far = perf_profiler_times.get(&category).unwrap_or(&zero);
+
+                so_far + duration
+            };
+
+            self.perf_profiler_times.borrow_mut().insert(category, aggregate);
+
+            const NANO: u64 = 1000 * 1000 * 1000;
+            const REPORT_INTERVAL: u64 = 10 * NANO;
+
+            match self.perf_profiler_next_report.get() {
+                None => self.perf_profiler_next_report.set(Some(start + REPORT_INTERVAL)),
+                Some(time) if time <= end => {
+                    self.perf_profiler_next_report.set(None);
+
+                    let stdout = stdout();
+                    let mut stdout = stdout.lock();
+                    writeln!(&mut stdout, "Script task time distribution:").unwrap();
+                    for (c, t) in self.perf_profiler_times.borrow().deref() {
+                        let secs = t / NANO;
+                        let nanos = t % NANO;
+                        writeln!(&mut stdout, "  {:?}: {}.{}s", c, secs, nanos).unwrap();
+                    }
+                    stdout.flush().unwrap();
+
+                    self.perf_profiler_times.borrow_mut().clear();
```

Comment:
Sorry, I didn't express myself clearly. I agree and implemented the change with a small modification. Instead of `start+REPORT_INTERVAL`, I'm using `end+REPORT_INTERVAL`.

I'm not sure what the correct workflow is. So far I've simply pushed more commits. Please tell me if/when I should squash any. Thanks. =)